rct229/rules/section5.py

```diff
@@ -110,6 +116,114 @@ def rule_check(self, context, calc_vals, data=None):
+        def __init__(self):
+            super(Section5Rule5.BuildingRule, self).__init__(
+                rmrs_used=UserBaselineProposedVals(False, True, False),
+                required_fields={},
+                each_rule=Section5Rule3.BuildingRule.RoofRule(),
+                index_rmr="baseline",
+            )
+
+        def create_context_list(context, data=None):
+            building = context.baseline
+            # List of all baseline roof surfaces to become the context for RoofRule
+            return [
+                UserBaselineProposedVals(None, surface, None)
+                for surface in find_all("$..surfaces[*]", building)
+                if get_opaque_surface_type(surface) == ROOF
+            ]
+
+        def create_data(self, context, data=None):
+            building = context.baseline
+            # Merge into the existing data dict
+            return {
+                **data,
+                surface_conditioning_category_dict: get_surface_conditioning_category_dict(
+                    data["climate_zone"], building
+                ),
+            }
+
+        class RoofRule(RuleDefinitionBase):
+            def __init__(self):
+                super(Section5Rule5.BuildingRule.RoofRule, self).__init__(
+                    rmrs_used=UserBaselineProposedVals(False, True, False),
+                    required_fields={},
+                )
+
+            def get_calc_vals(self, context, data=None):
+                climate_zone: str = data["climate_zone"]
+                scc: str = data["surface_conditioning_category_dict"][roof["id"]]
+                roof = context.baseline
+                roof_u_factor = roof["construction"]["u_factor"]
+
+                tartget_u_factor = None
+                target_u_factor_res = None
+                target_u_factor_nonres = None
+
+                if scc in [
+                    EXTERIOR_RESIDENTIAL,
+                    EXTERIOR_NON_RESIDENTIAL,
+                    SEMI_EXTERIOR,
+                ]:
+                    target_u_factor = table_G34_lookup(climate_zone, scc, ROOF)[
+                        "u_value"
+                    ]
+                elif scc == EXTERIOR_MIXED:
+                    target_u_factor_res = table_G34_lookup(
+                        climate_zone, EXTERIOR_RESIDENTIAL, ROOF
+                    )["u_value"]
+                    target_u_factor_nonres = table_G34_lookup(
+                        climate_zone, EXTERIOR_NON_RESIDENTIAL, ROOF
+                    )["u_value"]
+                    if target_u_factor_res == target_u_factor_nonres:
+                        target_u_factor = target_u_factor_res
+
+                return {
+                    "roof_u_factor": roof_u_factor,
+                    "target_u_factor": target_u_factor,
+                    "target_u_factor_res": target_u_factor_res,
+                    "target_u_factor_nonres": target_u_factor_nonres,
+                }
```

Comment:
Shall we add the ID of the surface in the calc_val?

Reply:
Each surface (roof) will be a node in the output which will include its id, the calc_vals, etc.  So I don't think we need to also add it to the calc_vals.